### PR TITLE
ntp: add controller (features #83)

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: src/libclient/controller/ntpcontroller.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+class NtpController::Private : public QObject
+{
+    Q_OBJECT
+
+public:
+    Private(NtpController *q)
+    : q(q)
+    {
+        connect(&timer, SIGNAL(timeout()), q, SLOT(update()));
+        timer.start(3600000);
```
Comment (MKV21) [inline src/libclient/controller/ntpcontroller.cpp]: Use a static const defined at the top of the file for the interval.

File: src/libclient/timing/ntp.h [hunk 2]
```diff
@@ -57,7 +59,7 @@ struct NtpPacket
     NtpTimestamp transmitTimestamp;
 };
 
-class Ntp : public QObject
+class CLIENT_API Ntp : public QObject
```
Comment (MKV21) [inline src/libclient/timing/ntp.h]: Can this class be deleted?

**Comment by MKV21** [comment]
Looks good, should we merge this now or wait for the other features? (local ntp servers, function to get synchronized time)

**Comment by monstermunchkin** [comment]
Regarding local time servers, maybe we should add this to the settings as I wouldn't know how to find them on the network. Does anyone else perhaps have an idea?

File: src/libclient/controller/ntpcontroller.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+void NtpController::readResponse()
+{
+    NtpPacket packet;
+    QHostAddress address;
+    quint16 port;
+
+    memset(&packet, 0, sizeof(packet));
+
+    if (m_socket->hasPendingDatagrams())
```
Comment (theoriginalgri) [inline src/libclient/controller/ntpcontroller.cpp]: Wouldn't it be more reliable to use a while loop in this case? If for some reason two packets are incoming and only one is read, the readyRead() signal might not be emitted again, so the controller would not work anymore.

**Comment by MKV21** [comment]
Another issue: The UdpSocket is opened once in the init function and the readyRead signal connected the whole time. Doesn't that mean every UDP packet sent to this port will be parsed by the readResponse function? This means even a port scan would set the network time (with garbage data?) as there is no checking of the received message. Maybe open the port just when sending and awaiting the first response and do some checking if the received data is a NtpPacket?

**Comment by monstermunchkin** [comment]
The UDP socket seems to be fine just like it is. I tested it and it didn't pick up any unwanted traffic.

**Comment by MKV21** [comment]
It does for me on the source port which I got via wireshark. Fortunately the app does not crash by this, it just uses a default value if it does not receive a valid time (1.1.1900). So it should be enough to open the socket only on sending a packet until the answer is received and checking if the answer is valid afterwards!

**Pull request merged**